src/System.Runtime.Extensions/src/System/AppDomain.cs

```diff
@@ -280,5 +286,45 @@ public void SetThreadPrincipal(IPrincipal principal)
                 _defaultPrincipal = principal;
             }
         }
+
+        public IPrincipal GetThreadPrincipal()
+        {
+            IPrincipal principal = _defaultPrincipal;
+            if (principal == null)
+            {
+                switch (_principalPolicy)
+                {
+                    case PrincipalPolicy.UnauthenticatedPrincipal:
+                        if (s_genericPrincipalCtor == null || s_genericIdentityCtor == null)
+                        {
+                            Assembly assembly = Assembly.Load(new AssemblyName("System.Security.Claims"));
+                            Type genericPrincipal = assembly.GetType("System.Security.Principal.GenericPrincipal", throwOnError: true);
+                            Type genericIdentity = assembly.GetType("System.Security.Principal.GenericIdentity", throwOnError: true);
+
+                            s_genericPrincipalCtor = genericPrincipal.GetConstructor(new[] { genericIdentity, typeof(string[]) });
+                            s_genericIdentityCtor = genericIdentity.GetConstructor(new[] { typeof(string), typeof(string) });
```

Comment:
Technically I think these should be done with Volatile.Write, since we're publishing shared state and want to make sure that writes while initializing these objects aren't moved to be later than the write that publishes.